client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java

```diff
@@ -532,66 +559,78 @@ private boolean moveToNextChunk() throws IOException {
+            }
 
-          currentChunk.readBytes(rawDataBuf, 0, size);
-        }
+            currentChunk.readBytes(compressedBuf, 0, size);
+          } else {
+            if (size > rawDataBuf.length) {
+              rawDataBuf = new byte[size];
+            }
 
-        // de-duplicate
-        if (attemptId == attempts[mapId]) {
-          if (!batchesRead.containsKey(mapId)) {
-            Set<Integer> batchSet = new HashSet<>();
-            batchesRead.put(mapId, batchSet);
+            currentChunk.readBytes(rawDataBuf, 0, size);
           }
-          Set<Integer> batchSet = batchesRead.get(mapId);
-          if (!batchSet.contains(batchId)) {
-            batchSet.add(batchId);
-            callback.incBytesRead(BATCH_HEADER_SIZE + size);
-            if (shuffleCompressionEnabled) {
-              // decompress data
-              int originalLength = decompressor.getOriginalLen(compressedBuf);
-              if (rawDataBuf.length < originalLength) {
-                rawDataBuf = new byte[originalLength];
+
+          // de-duplicate
+          if (attemptId == attempts[mapId]) {
+            if (!batchesRead.containsKey(mapId)) {
+              Set<Integer> batchSet = new HashSet<>();
+              batchesRead.put(mapId, batchSet);
+            }
+            Set<Integer> batchSet = batchesRead.get(mapId);
+            if (!batchSet.contains(batchId)) {
+              batchSet.add(batchId);
+              callback.incBytesRead(BATCH_HEADER_SIZE + size);
+              if (shuffleCompressionEnabled) {
+                // decompress data
+                int originalLength = decompressor.getOriginalLen(compressedBuf);
+                if (rawDataBuf.length < originalLength) {
+                  rawDataBuf = new byte[originalLength];
+                }
+                limit = decompressor.decompress(compressedBuf, rawDataBuf, 0);
+              } else {
+                limit = size;
               }
-              limit = decompressor.decompress(compressedBuf, rawDataBuf, 0);
+              position = 0;
+              hasData = true;
+              break;
             } else {
-              limit = size;
+              logger.debug(
+                  "Skip duplicated batch: mapId {}, attemptId {}, batchId {}.",
+                  mapId,
+                  attemptId,
+                  batchId);
             }
-            position = 0;
-            hasData = true;
-            break;
-          } else {
-            logger.debug(
-                "Skip duplicated batch: mapId {}, attemptId {}, batchId {}.",
-                mapId,
-                attemptId,
-                batchId);
           }
         }
-      }
 
-      return hasData;
+        return hasData;
+      } catch (IOException e) {
+        IOException ioe = e;
+        if (exceptionMaker != null) {
+          if (shuffleClient.reportShuffleFetchFailure(appShuffleId, shuffleId)) {
+            ioe =
+                new CelebornIOException(
```

Comment:
Not necessarily to throw FetchFailedException here, because FetchFailedException constructor automatically set fetchFailed
```
  // SPARK-19276. We set the fetch failure in the task context, so that even if there is user-code
  // which intercepts this exception (possibly wrapping it), the Executor can still tell there was
  // a fetch failure, and send the correct error msg back to the driver.  We wrap with an Option
  // because the TaskContext is not defined in some test cases.
  Option(TaskContext.get()).map(_.setFetchFailed(this))
```
, and Spark checks it even if the Exception doesn't match
```
        case t: Throwable if hasFetchFailure && !Utils.isFatalError(t) =>
          val reason = task.context.fetchFailed.get.toTaskFailedReason
          if (!t.isInstanceOf[FetchFailedException]) {
            // there was a fetch failure in the task, but some user code wrapped that exception
            // and threw something else.  Regardless, we treat it as a fetch failure.
            val fetchFailedCls = classOf[FetchFailedException].getName
            logWarning(s"TID ${taskId} encountered a ${fetchFailedCls} and " +
              s"failed, but the ${fetchFailedCls} was hidden by another " +
              s"exception.  Spark is handling this like a fetch failure and ignoring the " +
              s"other exception: $t")
          }
          setTaskFinishedAndClearInterruptStatus()
          execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
```

Reply:
Agree, but we should add this comment to the code - it can get confusing for future code evolution otherwise :-)
Something like:

```
/*
 * [[ExceptionMaker.makeException]], for spark applications with celeborn.client.spark.fetch.throwsFetchFailure enabled will result in creating 
 * a FetchiFailedException; and that will make the TaskContext as failed with shuffle fetch issues - see SPARK-19276 for more.
 * Given this, Celeborn can wrap the FetchFailedException with our CelebornIOException
*/

```

Reply:
Done, thanks